liblcm-test/lcm-source.1
.TH lcm_source 1 2007-12-13 "LCM" "Lightweight Communications and Marshalling (LCM)"
.SH NAME
lcm_source \- transmit random messages over LC channels
.SH SYNOPSIS
.TP 5
\fBlcm_source \fI[options]\fR

.SH DESCRIPTION
.PP
\fBlcm_source\fR is a utility for testing the Lightweight Communications library
(LC).  \fBlcm_source\fR transmits random messages on either a specified LC
channel or random LC channels.  Messages can be randomly sized, or a specified
fixed size.

.SH OPTIONS
The following options are provided by \fBlcm_source\fR:
.TP
.B \-m \fICHAN\fR
Transmits messages on channel \fICHAN\fR (default: random channels)
.TP
.B \-p \fIDELAY\fR
Wait \fIDELAY\fR microseconds between transmissions (default: 10000)
.TP
.B \-s \fISIZE\fR
Transmit packets of size \fISIZE\fR bytes.  If \fISIZE\fR is 0 (default),
then packets are randomly sized between 1 and 65,000 bytes
.TP
.B \-v
verbose mode.  Prints a summary of each packet
.TP
.B \-h
Prints this help text and exits.

.SH COPYRIGHT

lcm-source is part of the Lightweight Communications and Marshalling (LCM) project.
Permission is granted to copy, distribute and/or modify it under the terms of
the GNU Lesser General Public License as published by the Free Software
Foundation; either version 2.1 of the License, or (at your option) any later
version.  See the file COPYING in the LCM distribution for more details
regarding distribution.

LCM is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.
You should have received a copy of the GNU Lesser General Public
License along with LCM; if not, write to the Free Software Foundation, Inc., 51
Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

.SH AUTHOR

This manual page was written by Albert Huang.


NEWS
August 30, 2018
==============

Release 1.4.0

Significant changes:

- Build system
  - Switch to CMake
  - Add CPack support
- lcm-gen
  - Allow arrays to be sized with const values
- C
  - Add lcm_subscription_get_queue_size()
  - Standardize C/C++ code formatting with clang-format
- C++
  - Add support for some C++11 features (lambda subscribe)
  - add lcm::Subscription::getQueueSize()
- Go
  - Experimental support for golang
- Python
 - Add __dimensions__ and __typenames__ fields for better introspection
- lcm-logger
  - flush buffers before exiting
- General
  - Lots of misc bugfixes

April 29, 2016
==============

Release 1.3.1

This is primarily a bugfix and maintenance release.

- Java
  - Fix lcm-java automake for out of source builds
- Lua
  - Add support for Lua 5.3
- Python
  - Fix python LCM code generation when a package is not specified
- General
  - Minor documentation updates

Oct 28, 2015
============

Release 1.3.0

This is primarily a bugfix and maintenance release.

- C# / .NET
  - fix lcm-dotnet Close (closes issue #18)
- Windows
  - Add socket window size notes in documentation
  - Visual Studio build fixes, update to VS2013
  - Cygwin build fixes
- C / C++
  - Reject invalid multicast addresses
  - Use last fragment timestamp instead of first for message timestamping
  - binary search bugfix in lcm_eventlog_seek_to_timestamp()
  - API changes:
    - add FILE* LogFile::getFilePtr()
    - void LCM::unbsubscribe() -> int LCM::unsubscribe()
- Python
  - Python 3 compatibility bugfixes
  - setup.py use subprocess instead of deprecated commands package
  - EventLog.seek() use fseeko() not fseek()
- lcm-spy
  - misc bugfixes
  - fix bitshifting of signed integer types
- lcm-gen
  - add --cpp-std option for C++11 support in generated types.


Jan 21, 2015
============

Release 1.2.1

This is a bugfix and maintenance release.

- Documentation updates
- lcm-spy
   - Misc bugfixes
   - Fix bug where lcm-spy did not display content of some messages (Johan
     Widen)
- Remove jchart2d jar file from INSTALLED_CLASSPATH variable

Nov 18, 2014
============

Release 1.2.0

This release adds charting to lcm-spy (Andy Barry) and updates the lcm-spy GUI.
This release also marks the official migration of LCM to GitHub.  The new site
is https://github.com/lcm-proj/lcm

- C / C++
    - Use select() instead of poll() in lcm_mpudpm.c
- lcm-spy
    - Add sparklines and charting (Andy Barry), via JChart2d
- Python
    - Fix setup.py
- Java
    - TCPService convert mutex to read/write lock.  Reduces deadlock when
      send/receive buffers are full.
    - Make TCPService interruptible (closes googlecode issue #94)
- lcm-gen
    - Fix regression for comments appearing in various places in a type
      definition file (closes googlecode issue #97)
- misc
    - Build system cleanups
    - Add more unit tests, cleanup existing tests.
    - Documentation fixes.
- OS/X
    - reduce max packet sizes (closes googlecode issue #99)
    - fix test/cpp/Makefile

Aug 19, 2014
============

Release 1.1.2

This is a bugfix and maintenance release.

- C / C++
    - memset struct msghdr to 0 before calling recvmsg (fixes issue #90).
    - fix eventlog memory leak when errors occur.
- lcm-gen
    - Apply patch from bibocc to fix lcm-gen regression (fixes issue #92).
    - fix emit-cpp for the case where it needs to includ both `<vector>` and
    `<string>`.
    - Resize vector in decode after checking whether the size > 0.
    - fix tokenizer regression in parsing empty comment blocks (fixes issue #93).
    - Add --version flag.
    - Update manpage.

Jul 29, 2014
============

Release 1.1.1

This is a bugfix and maintenance release.

- Python
    - Fix multithreading issue
- lcm-gen
    - C / C++ bindings include field comments in generated files.

Jul 16, 2014
============

Release 1.1.0

- General
    - Add "memq://" provider to support unit testing code that uses LCM.
    Supported in C, C++, Python, lua, Java.
- C
    - lcm_tcpq.c add missing `#include <signal.h>`
    - Add introspection to generated C types
    - Increase fragmentation threshold.
    - Add lcm_handle_timeout() function
- C++
    - Fix emit_cpp for messages that contain a single non-primitive type
    - Declare const fields using enums instead of static const int
    - Some const correctness
    - Add lcm::LCM::handleTimeout() function
- Python
    - Add Python3 compatibility
    - Add lcm.LCM.handle_timeout() function
    - Sub-types are initialized to valid instances instead of None
- lua
    - Add handle_timeout() function, deprecate timedhandle()
- lcm-logger
    - Remove --auto-split-hours option
    - Rename --auto-split-mb option to --split-mb
    - Add --rotate option for rotating log files
    - Add --quiet option
- lcm-gen
    - Add --package-prefix option
    - Detect duplicate types in input.
    - C emit_hash_recursive function explicit cast result to int64_t (closes #81)
    - Detect corrupt log files.
- lcm-spy
    - Better drawing of long type names
- lcm-logplayer
    - Actually accept -l/--lcm-url optin
- misc
    - Add more unit tests

Jun 8, 2013
===========

Release 1.0.0

This release adds experimental support for Lua (contributed by Tim Perkins)

- C
    - tcpq provider
        - unsubscribe correctly send unsubscribe request to server, not subscribe.
        - automatically reconnect when possible.
    - fix lcm_internal_pipe in windows - don't use addrlen param in accept()
    - lcm_coretypes use void casting to avoid unused parameter warning instead of
      unnamed parameter.
    - Cast to struct timeval and not GTimeVal when getting packet timestamp.
- Windows
    - Ignore the return value of setsockopt() when joining the mc group for the
      send fd.
- C# / .NET
    - replace Dns.GetHostEntry by Dns.GetHostAddresses - should solve problems
    with reverse DNS lookups
- lcm-gen
    - C
        - avoid unused parameter warnings in generated types
    - C#

liblcm-test/lcm-source.c
// file: lcm_source.c
// desc: utility to basically spew garbage on the LC channels to use up
//       bandwidth.

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef WIN32
#include <winsock2.h>
#endif

#ifndef WIN32
#include <signal.h>
#endif

#include <getopt.h>
#include <lcm/lcm.h>
#include <time.h>

#define DEFAULT_TRANSMIT_INTERVAL_USEC 10000

static int g_quit = 0;

static void make_msg_channel(char *buf, int maxlen)
{
    char *cset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    int i;
    int ind;
    int nchars = strlen(cset);
    if (0 == maxlen)
        return;

    for (i = 0; i < maxlen - 1; i++) {
        ind = (int) (nchars * ((double) rand() / (RAND_MAX + 1.0)));
        buf[i] = cset[ind];
    }
    buf[maxlen - 1] = 0;
}

static void usage()
{
    printf(
        "usage: lcm_source [OPTIONS]\n"
        "\n"
        "periodically transmits LC messages.\n"
        "\n"
        "  -h     prints this help text and exits\n"
        "  -m s   transmits messages on channel s (otherwise random channel)\n"
        "  -p t   wait t microseconds between transmissions (default %d)\n"
        "  -s n   transmit packets of size n bytes.  If n is 0 (default),\n"
        "         then packets are randomly sized between 1 and 65,000\n"
        "         bytes\n"
        "  -v     verbose mode. Prints a summary of each packet\n",
        DEFAULT_TRANSMIT_INTERVAL_USEC);
    exit(1);
}

#ifndef WIN32
void on_signal(int signum)
{
    g_quit = 1;
}
#endif

int main(int argc, char **argv)
{
    int verbose = 0;
    int random_channel = 1;
    char channel[80];
    uint8_t *data = NULL;
    int transmit_interval_usec = DEFAULT_TRANSMIT_INTERVAL_USEC;
    int packet_sz = 0;
    // memset (data, 0, sizeof(data));

    char *optstring = "hm:p:s:v";
    int c;

#ifndef WIN32
    signal(SIGINT, on_signal);
#endif

    while ((c = getopt(argc, argv, optstring)) >= 0) {
        switch (c) {
        case 'h':
            usage();
            break;
        case 'm':
            random_channel = 0;
            strncpy(channel, optarg, sizeof(channel));
            break;
        case 'p':
            transmit_interval_usec = atoi(optarg);
            if (transmit_interval_usec < 0) {
                usage();
            }
            break;
        case 's':
            packet_sz = atoi(optarg);
            if (packet_sz < 0)
                usage();
            break;
        case 'v':
            verbose = 1;
            break;
        default:
            usage();
            break;
        };
    }

    if (0 == packet_sz) {
        data = (uint8_t *) malloc(65536);
    } else {
        data = (uint8_t *) malloc(packet_sz);
        memset(data, 0, packet_sz);
    }

    lcm_t *lcm = lcm_create(NULL);
    if (!lcm) {
        fprintf(stderr, "couldn't allocate lcm_t\n");
        return 1;
    }

    srand(time(NULL));

    uint32_t seqno = 0;
    while (!g_quit) {
        int sz = 0;

        // pick a random message type if no fixed type was specified...
        if (random_channel) {
            make_msg_channel(channel, 10);
        }

        // pick a random or fixed packet size (depending on cmd-line option)
        if (0 == packet_sz) {
            sz = 1 + (int) (65000 * ((double) rand() / (RAND_MAX + 1.0)));
        } else {
            sz = packet_sz;
        }

        if (sz >= sizeof(seqno))
            memcpy(data, &seqno, sizeof(seqno));

        // spew
        lcm_publish(lcm, channel, data, sz);

        if (verbose) {
            printf("packet type: [%s] size: %d\n", channel, sz);
        }

        if (transmit_interval_usec > 0) {
#ifdef WIN32
            // sleep...
            Sleep(transmit_interval_usec / 1000);
#else
            struct timespec ts = {.tv_sec = (int) (transmit_interval_usec / 1000000),
                                  .tv_nsec = (transmit_interval_usec % 1000000) * 1000};
            if (0 != nanosleep(&ts, NULL)) {
                perror("nanosleep");
            }
#endif
        }
        seqno++;
    }
    printf("Transmitted %d messages\n", seqno);
    free(data);

    lcm_destroy(lcm);

    return 0;
}
